chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/job/summarystatistics/AsynchronousSummaryStatistics.java

```diff
@@ -365,204 +425,471 @@
+
+        List<SurveillanceNonconformityEntity> nonconformitiesWithDeterminationDate = surveillances.stream()
+                .flatMap(surv -> surv.getSurveilledRequirements().stream())
+                .flatMap(req -> req.getNonconformities().stream())
+                .filter(nc -> nc.getDateOfDetermination() != null)
+                .distinct()
+                .collect(Collectors.toList());
+
+        Long totalDuration = nonconformitiesWithDeterminationDate.stream()
+                .map(nc -> getDaysToAssessNonconformtity(findSurveillanceForNonconformity(nc, surveillances), nc))
+                .collect(Collectors.summingLong(n -> n.longValue()));
+
+        return new AsyncResult<Long>(totalDuration / nonconformitiesWithDeterminationDate.size());
+    }
+
+    @Async("jobAsyncDataExecutor")
+    @Transactional
+    public Future<Long> getAverageTimeToApproveCAP(SurveillanceStatisticsDAO surveillanceStatisticsDAO) {
+        List<SurveillanceEntity> surveillances = surveillanceStatisticsDAO.getAllSurveillancesWithNonconformities();
+
+        List<SurveillanceNonconformityEntity> nonconformities = surveillances.stream()
+                .flatMap(surv -> surv.getSurveilledRequirements().stream())
+                .flatMap(req -> req.getNonconformities().stream())
+                .filter(nc -> nc.getDateOfDetermination() != null && nc.getCapApproval() != null)
+                .distinct() // Want to figure out how to get rid of this
+                .collect(Collectors.toList());
+
+        Long totalDuration = nonconformities.stream()
+                .map(nc -> Math
+                        .abs(ChronoUnit.DAYS.between(nc.getDateOfDetermination().toInstant(), nc.getCapApproval().toInstant())))
+                .collect(Collectors.summingLong(n -> n.longValue()));
+
+        return new AsyncResult<Long>(totalDuration / nonconformities.size());
+    }
+
+    @Async("jobAsyncDataExecutor")
+    @Transactional
+    public Future<Long> getAverageDurationOfCAP(SurveillanceStatisticsDAO surveillanceStatisticsDAO) {
+        List<SurveillanceEntity> surveillances = surveillanceStatisticsDAO.getAllSurveillancesWithNonconformities();
+
+        List<SurveillanceNonconformityEntity> nonconformities = surveillances.stream()
+                .flatMap(surv -> surv.getSurveilledRequirements().stream())
+                .flatMap(req -> req.getNonconformities().stream())
+                .filter(nc -> nc.getCapApproval() != null)
+                .distinct() // Want to figure out how to get rid of this
+                .collect(Collectors.toList());
+
+        Long totalDuration = nonconformities.stream()
+                .map(nc -> Math
+                        .abs(ChronoUnit.DAYS.between(
+                                nc.getCapApproval().toInstant(),
+                                nc.getCapEndDate() != null ? nc.getCapEndDate().toInstant() : Instant.now())))
```

Comment:
What happens if capEndDate is null? In the method "getAverageTimeToCloseSurveillance" you specifically filter down to where an end date is not null so wondering if that should have been done above in the filtering stage of this method as well?

Reply:
If the capEndDate is null we use the current date.  
Here is the requirement:
_Numerator:  Total number of days from CAP_Approval_Date to (Was_Complete_Date or today)_